cpp/src/arrow/python/builtin_convert.cc

```diff
@@ -507,7 +552,13 @@ Status ConvertPySequence(PyObject* obj, MemoryPool* pool, std::shared_ptr<Array>
   std::shared_ptr<DataType> type;
   int64_t size;
   PyDateTime_IMPORT;
-  RETURN_NOT_OK(InferArrowType(obj, &size, &type));
+  RETURN_NOT_OK(InferArrowTypeAndSize(obj, &size, &type));
+  return ConvertPySequence(obj, pool, out, type, size);
+}
+
+Status ConvertPySequence(PyObject* obj, MemoryPool* pool, std::shared_ptr<Array>* out,
+    const std::shared_ptr<DataType>& type, int64_t size) {
+  PyDateTime_IMPORT;
```

Comment:
How about putting this in

https://github.com/apache/arrow/blob/master/cpp/src/arrow/python/config.cc#L25

`Init()` gets calls when you import pyarrow